File: Multiple_linear_regression.R
## Multiple Linear Regression ##

# Remove all objects from the workspace (memory) of the R session
rm(list=ls())
gc()

# Import libraries
# install.packages('caTools')
# install.packages('ggplot2')
library(caTools)
library(ggplot2)

# Set working directory
setwd('C:/Machine-Learning-A-Z-course/Machine Learning A-Z (Codes and Datasets)/Part 2 - Regression/Section 5 - Multiple Linear Regression/R')
getwd()

# Importing the data set
data_startup <- read.csv("50_Startups.csv")


# Encoding Categorical data
data_startup$State <- factor(data_startup$State,
                          levels = c('New York', 'California', 'Florida'),
                          labels = c(1,2,3))


# Splitting the data set into the training set and test set
set.seed(123)
split <- sample.split(data_startup$Profit,
                      SplitRatio = 0.8) # percentage of the training set so 80% 
# Create the training and test set
training_set <- subset(data_startup, split == TRUE)
test_set <- subset(data_startup, split == FALSE)

# Fitting multiple linear regression to the training set
# . means all independent variables
regressor <- lm(formula = Profit ~ .,
                data = training_set)


summary(regressor)

# Predicting the Test set result
y_pred <- predict(regressor, newdata = test_set)

# Backward elimination to predict better the model
# Eliminate variable if the p values is higher than the Significant value 0.05-5%
names(data_startup)
regressor <- lm(formula = Profit ~ R.D.Spend,
                data = data_startup)
summary(regressor)

# eliminated <- + State
# eliminated <- + Administration
# eliminated <-  + Marketing.Spend

# Function to perform the backward Elimination
backwardElimination <- function(x, sl){
  numVars <-  length(x)
  for (i in c(1:numVars)){
    regressor <-  lm(formula = Profit ~ ., data = x)
    maxVar  <- max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j <-  which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x <- x[, -j]
    }
    numVars = numVars -1
  }
  return(summary((regressor)))
}


# sl <- 0.05
# dataset <- dataset[, c(1,2,3,4,5)]

backwardElimination(training_set, sl)

File: Multiple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 19 10:10:17 2024

@author: no67wuwu
wrok: Multiple linear regresion

No need to apply the Feature scaling 

"""
# import the libraries
import pandas as pd
import numpy as np

# Import the data set
data_startup = pd.read_csv(r"C:\Machine-Learning-A-Z-course\Machine Learning A-Z (Codes and Datasets)\Part 2 - Regression\Section 5 - Multiple Linear Regression\Python\50_Startups.csv")

x = data_startup.iloc[:, :-1].values
y = data_startup.iloc[:, -1].values # Select the dependet variable 


# Encoding the categorical variable in this case is the State
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [3])],
                       remainder='passthrough') # if there is not passth does not go throu rows
X = np.array(ct.fit_transform(x)) # create the array with the transormed. The column country has been trasformed to 01 for example France is coding 1 0 0 each nunber is a column
print(X)


## Splitting the dataser into tha Training set and Test set
from sklearn.model_selection import train_test_split
# we define the training data ans test data with the function train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size= 0.2, random_state = 0)

print(x_train)
print(x_test)
print(y_train)
print(y_test )


# Trining the Multiple Linear Regression model on the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()

# With the fit method we train our model
regressor.fit(x_train, y_train)

# Predict the Test set result
y_pred = regressor.predict(x_test) # we added just the year of Experience because we want to predic the salary
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))


























